server/src/main/java/org/cloudfoundry/identity/uaa/provider/token/ClientAssertionHeaderAuthenticator.java

```diff
@@ -0,0 +1,33 @@
+package org.cloudfoundry.identity.uaa.provider.token;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.jwt.Jwt;
+import org.springframework.security.jwt.JwtHelper;
+import org.springframework.security.jwt.crypto.sign.RsaVerifier;
+import org.springframework.util.StringUtils;
+
+public class ClientAssertionHeaderAuthenticator {
+    private static final Log logger = LogFactory.getLog(ClientAssertionHeaderAuthenticator.class);
+
+    /**
+     * @param jwtToken  with 'sub' and 'iss' claims for the client 
```

Comment:
actually this will need 'tenant_id' and 'sub' claims , which are needed to lookup device's public key.